File: skellig-test-step-processing-rmq/src/main/java/org/skellig/teststep/processor/rmq/RmqTestStepProcessor.java
package org.skellig.teststep.processor.rmq;

import com.typesafe.config.Config;
import org.skellig.teststep.processing.converter.TestStepResultConverter;
import org.skellig.teststep.processing.processor.BaseTestStepProcessor;
import org.skellig.teststep.processing.processor.TestStepProcessor;
import org.skellig.teststep.processing.state.TestScenarioState;
import org.skellig.teststep.processing.validation.TestStepResultValidator;
import org.skellig.teststep.processor.rmq.model.RmqDetails;
import org.skellig.teststep.processor.rmq.model.RmqTestStep;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class RmqTestStepProcessor extends BaseTestStepProcessor<RmqTestStep> {

    private Map<String, RmqChannel> rmqChannels;

    protected RmqTestStepProcessor(Map<String, RmqChannel> rmqChannels,
                                   TestScenarioState testScenarioState,
                                   TestStepResultValidator validator,
                                   TestStepResultConverter testStepResultConverter) {
        super(testScenarioState, validator, testStepResultConverter);
        this.rmqChannels = rmqChannels;
    }

    @Override
    protected Object processTestStep(RmqTestStep testStep) {
        Object response = null;
        Optional<String> sendTo = testStep.getSendTo();
        Optional<String> receiveFrom = testStep.getReceiveFrom();
        String routingKey = testStep.getRoutingKey();

        sendTo.ifPresent(channelId -> send(testStep.getTestData(), channelId, routingKey));

        if (receiveFrom.isPresent()) {
            RmqChannel channel = rmqChannels.get(receiveFrom.get());
            Optional<String> respondTo = testStep.getRespondTo();
            Object responseTestData = testStep.getTestData();
            response = channel.read(respondTo.isPresent() ? null : responseTestData, testStep.getTimeout());

            if (respondTo.isPresent() && response != null) {
                send(responseTestData, respondTo.get(), routingKey);
            }
        }
        return response;
    }

    private void send(Object testData, String channelId, String routingKey) {
        RmqChannel channel = rmqChannels.get(channelId);
        channel.send(testData, routingKey);
    }

    @Override
    public void close() {
        rmqChannels.values().forEach(RmqChannel::close);
    }

    @Override
    public Class<RmqTestStep> getTestStepClass() {
        return RmqTestStep.class;
    }

    public static class Builder extends BaseTestStepProcessor.Builder<RmqTestStep> {

        private Map<String, RmqChannel> rmqChannels;
        private RmqDetailsConfigReader rmqDetailsConfigReader;

        public Builder() {
            rmqDetailsConfigReader = new RmqDetailsConfigReader();
            rmqChannels = new HashMap<>();
        }

        public Builder withRmqChannel(RmqDetails rmqDetails) {
            this.rmqChannels.putIfAbsent(rmqDetails.getChannelId(), new RmqChannel(rmqDetails));
            return this;
        }

        public Builder withRmqChannels(Config config) {
            rmqDetailsConfigReader.read(config).forEach(this::withRmqChannel);
            return this;
        }

        @Override
        public TestStepProcessor<RmqTestStep> build() {
            return new RmqTestStepProcessor(rmqChannels, testScenarioState, validator, testStepResultConverter);
        }
    }
}

File: skellig-test-step-processing-rmq/src/main/kotlin/org/skellig/teststep/processor/rmq/RmqTestStepProcessor.kt
package org.skellig.teststep.processor.rmq

import com.rabbitmq.client.AMQP
import org.skellig.teststep.processing.processor.BaseTestStepProcessor
import org.skellig.teststep.processing.processor.TestStepProcessor
import org.skellig.teststep.processing.state.TestScenarioState
import org.skellig.teststep.processing.util.debug
import org.skellig.teststep.processing.util.info
import org.skellig.teststep.processing.util.logger
import org.skellig.teststep.processor.rmq.model.RmqTestStep

/**
 * RmqTestStepProcessor is a class that processes RMQ test steps.
 * It processes the test step this way:
 * 1) Sends a message ([RmqTestStep.testData]) to [RmqTestStep.sendTo]
 * 2) Reads a message ([RmqTestStep.testData]) from [RmqTestStep.readFrom]. If [RmqTestStep.respondTo] defined, then
 * responds with message ([RmqTestStep.testData]) to [RmqTestStep.respondTo] if the received message is valid
 * based on [RmqTestStep.validationDetails].
 *
 * All send, read and respond operations are performed in parallel per each channel but blocked until they are all finished.
 *
 * @param rmqChannels A map of RMQ channels where the key is the channel ID and the value is the [RmqChannel] instance.
 * @param testScenarioState The state of the test scenario.
 */
open class RmqTestStepProcessor(
    protected val rmqChannels: Map<String, RmqChannel>,
    testScenarioState: TestScenarioState?
) : BaseTestStepProcessor<RmqTestStep>(testScenarioState!!) {

    private val log = logger<RmqTestStepProcessor>()

    override fun processTestStep(testStep: RmqTestStep): Any? {
        var response: Map<*, Any?>? = null
        val sendTo = testStep.sendTo
        val readFrom = testStep.readFrom
        val routingKey = testStep.routingKey

        sendTo?.let {
            log.info(testStep, "Start to send message of test step '${testStep.name}' to RMQ queues $sendTo")
            send(testStep, it, routingKey, testStep.getAmqpProperties())
        }

        readFrom?.let {
            log.info(testStep, "Start to read message of test step '${testStep.name}' from RMQ queues $readFrom")
            val respondTo = testStep.respondTo
            response = read(testStep, readFrom, if (respondTo != null) null else testStep.testData)
            respondTo?.let {
                if (isValid(testStep, response)) {
                    log.info(testStep, "Respond to received message to RMQ queues '$respondTo'")
                    send(testStep, respondTo, routingKey, testStep.getAmqpProperties())
                } else {
                    log.warn("Can't respond to '{}' queues because the response is not valid: {}", respondTo, response)
                }
            }
        }
        return response
    }

    private fun read(testStep: RmqTestStep, channels: Set<String?>, responseTestData: Any? = null): Map<*, Any?> {
        val tasks = channels.associateWith {
            {
                val channel = rmqChannels[it] ?: error(getChannelNotExistErrorMessage(it))
                log.debug(testStep) { "Start to read message from RMQ queue '$it'" }
                val message = channel.read(responseTestData)
                log.debug(testStep) { "Received message from RMQ queue '$it'" }
                message
            }
        }
        return runTasksAsyncAndWait(tasks, testStep)
    }

    private fun send(testStep: RmqTestStep, channels: Set<String>, routingKey: String?, properties: AMQP.BasicProperties?) {
        val tasks = channels.associateWith {
            {
                val channel = rmqChannels[it] ?: error(getChannelNotExistErrorMessage(it))
                log.debug(testStep) { "Send message to RMQ queue '$it'" }
                channel.send(testStep.testData, routingKey, properties)
                log.debug(testStep) { "Message has sent to RMQ queue '$it'" }
                "sent"
            }
        }
        runTasksAsyncAndWait(tasks, testStep)
    }

    override fun close() {
        log.info("Close RMQ Test Step Processor and all connections to queues")
        rmqChannels.values.forEach { it.close() }
    }

    override fun getTestStepClass(): Class<RmqTestStep> {
        return RmqTestStep::class.java
    }

    private fun getChannelNotExistErrorMessage(channelId: String?) =
        "Channel '$channelId' was not registered in RMQ Test Step Processor"

    class Builder : BaseRmqProcessorBuilder<RmqTestStep>() {
        override fun build(): TestStepProcessor<RmqTestStep> {
            return RmqTestStepProcessor(rmqChannels, testScenarioState)
        }
    }
}